<a href="https://colab.research.google.com/github/thejkiran/My_Projects/blob/main/diabetes_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing required libraries

In [3]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score

Importing dataset and analysis

In [4]:
#loading the PIMa dataset which contains only data of pregnant females
df = pd.read_csv('/content/diabetes.csv')

In [5]:
df.head()

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1

[5 rows x 9 columns]

In [6]:
df.describe()

Pregnancies     Glucose  ...         Age     Outcome
count   768.000000  768.000000  ...  768.000000  768.000000
mean      3.845052  120.894531  ...   33.240885    0.348958
std       3.369578   31.972618  ...   11.760232    0.476951
min       0.000000    0.000000  ...   21.000000    0.000000
25%       1.000000   99.000000  ...   24.000000    0.000000
50%       3.000000  117.000000  ...   29.000000    0.000000
75%       6.000000  140.250000  ...   41.000000    1.000000
max      17.000000  199.000000  ...   81.000000    1.000000

[8 rows x 9 columns]

In [7]:
df.shape

(768, 9)

So there are 768 rows and 9 columns, which means we have 768 examples.
Out of 9 columns Outcome column is our output variable which explains 1 as Diabetic and 0 as Non - diabteic

In [8]:
#Lets count number of diabetic(1) and number of non - diabetic patients(0)
df['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

So here we come to know that 500 examples are of non - diabetic and 268 examples of diabetic

In [9]:
#Analysis of output variable
df.groupby('Outcome').mean()

Pregnancies     Glucose  ...  DiabetesPedigreeFunction        Age
Outcome                           ...                                     
0           3.298000  109.980000  ...                  0.429734  31.190000
1           4.865672  141.257463  ...                  0.550500  37.067164

[2 rows x 8 columns]

In [10]:
#Lets assign the variables to dependant and independant variables to seperate output from the dataset
X = df.drop(columns='Outcome', axis=1)

#Here axis = 1 is given so as to drop column, whereas axis = 0 drops specified rows

Y = df['Outcome']

In [11]:
X.head()

Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0            6      148             72  ...  33.6                     0.627   50
1            1       85             66  ...  26.6                     0.351   31
2            8      183             64  ...  23.3                     0.672   32
3            1       89             66  ...  28.1                     0.167   21
4            0      137             40  ...  43.1                     2.288   33

[5 rows x 8 columns]

Lets standardize the data so that all the values will be ranged from -1 to +1 so that machine can understand to its understanding

In [12]:
#lets import standardscaler function
Scaler = StandardScaler()

In [13]:
Scaler.fit(X)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [14]:
Standard_Scaler = Scaler.transform(X)

In [15]:
# we can do both fit and tronsform in a single step as
#Standard_Scaler = Scaler.fit_transform(X)

In [16]:
print(Standard_Scaler)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [17]:
#Assign the variables again
X = Standard_Scaler
Y = df['Outcome']

In [18]:
Y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

Lets Split the data into train and test data
for example a student preparing for maths exam can be considered train data and the questions he is going to face in exam will be the test data

In [19]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y , test_size = 0.2, stratify = Y , random_state = 2)

Here test size = 0.2 mans 20% of the data will be taken as test and 80% of data will be used to train.
Straitify is used to declare output variable and it splits the output data in equal proportions.

In [20]:
#lets check the original data with train and test data
print(X.shape, X_train.shape, X_test.shape)

(768, 8) (614, 8) (154, 8)


Lets train the data
importing svm classifier

In [21]:
classifier = svm.SVC(kernel = 'linear')

In [22]:
classifier.fit(X_train, Y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [23]:
#lets check the accuracy score on training data
X_train_prediction = classifier.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [24]:
#printing the accuracy
print("The accuracy of training data is : ", training_data_accuracy)

The accuracy of training data is :  0.7866449511400652


In [25]:
#lets check the accuracy score on testing data
X_test_prediction = classifier.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [26]:
#printing the accuracy
print("The accuracy of testing data is : ", testing_data_accuracy)

The accuracy of testing data is :  0.7727272727272727


In [32]:
#testing the data
input_data = (1,109,56,21,135,25.2,0.833,23)

#input data as array as we standardized the data
Input_data_as_array = np.asarray(input_data)

#reshaping the data for one instance
input_data_reshaped = Input_data_as_array.reshape(1,-1)

#standardize the data
test_input = Scaler.transform(input_data_reshaped)

predict = classifier.predict(test_input)
print(predict)

if (predict[0] == 0):
  print("The person is Non - Diabetic")
else:
  print("The person is Diabetic")

[0]
The person is Non - Diabetic
